### PR TITLE
Support workspace cloaking through a delimited list of server paths

### PR DESCRIPTION
**Pull request by ajalexander**
This change adds support for cloaking in the TFS workspaces. Along with providing the server path, a delimited list of server paths can optionally be included for cloaking. During workspace creation, each of the cloak paths are cloaked after the workspace is initially mapped.

This should resolve [JENKINS-4709](https://issues.jenkins-ci.org/browse/JENKINS-4709).

For those wanting to specify multiple paths without changing the folder name or relative layout of the files, this also handles [JENKINS-4542](https://issues.jenkins-ci.org/browse/JENKINS-4542) and [JENKINS-11007](https://issues.jenkins-ci.org/browse/JENKINS-11007).

### CODE REVIEW

**Comment by mrcaron** [comment]
I'm very interested to see this PR come through. It's a big item for my team and I'm using @ajalexander's repo until the jenkins team integrates his changes.

**Comment by olivierdagenais** [comment]
Sorry for the very long delay; when I first got commit access to this project, I had been only tackling my immediate needs at $DAYJOB.

I am planning to review these changes for release 3.3.0.

**Comment by mrcaron** [comment]
Awesome!

**Comment by mmitche** [comment]
Any plans on getting this merged?  This is a big request for my team too :)

**Comment by olivierdagenais** [comment]
Yes, all in good time. Please stand by.

**Comment by mmitche** [comment]
It seems that it might be better to have this be a more general mapping feature (selective cloaking and syncing).  The issue here is that in really large repositories, it's a lot easier to start with "I need X, Y and Z" in general, but "don't sync X\A or Y\B".

I may end up having to do this feature for the Jenkins instance I'm bringing up.  I may lay most of the work on top of this sem's PR.

**Comment by ajalexander** [comment]
@mmitche Are you proposing a different (or additional) way to configure what is mapped while still using the same cloaking mechanism behind the scene? Or did you have a different idea of how you planned to setup the workspace mapping?

**Comment by mmitche** [comment]
I think I'd like to have a general workspace mapping (would require changes to the cloaking code in the PR) so the synced folders were defined more closely to the way the TFS defines them (sync X, Y and Z but not Z/Y or X/Q), rather than syncing everything under a project (which is just a TFS folder) except for what we cloak.

The issue is that for really big projects, the list of things cloaked can be much larger than a list of  selective maps and cloaks.

**Comment by mahsoud** [comment]
@ajalexander this change works great in our environment. Thank you.

File: src/main/java/hudson/plugins/tfs/TeamFoundationServerScm.java [hunk 1]
```diff
@@ -178,7 +197,7 @@ private String substituteBuildParameter(Run<?,?> run, String text) {
     public boolean checkout(AbstractBuild<?, ?> build, Launcher launcher, FilePath workspaceFilePath, BuildListener listener, File changelogFile) throws IOException, InterruptedException {
         Server server = createServer(new TfTool(getDescriptor().getTfExecutable(), launcher, listener, workspaceFilePath), build);
         try {
-            WorkspaceConfiguration workspaceConfiguration = new WorkspaceConfiguration(server.getUrl(), getWorkspaceName(build, Computer.currentComputer()), getProjectPath(build), getLocalPath());
+            WorkspaceConfiguration workspaceConfiguration = new WorkspaceConfiguration(server.getUrl(), getWorkspaceName(build, Computer.currentComputer()), getProjectPath(build), cloakPaths, getLocalPath());
```
Comment (mmitche) [inline src/main/java/hudson/plugins/tfs/TeamFoundationServerScm.java]: cloakPaths here should have its build variables replaced.

**Comment by ajalexander** [comment]
This PR has been obsoleted by PR #60 (which uses the Java SDK to accomplish the same goal).

File: src/main/java/hudson/plugins/tfs/TeamFoundationServerScm.java (hunk 1 above)
Comment (ajalexander) [inline src/main/java/hudson/plugins/tfs/TeamFoundationServerScm.java]: The new version of this (PR #60) does the build variable replacement.

**Pull request closed without merging**